baseImages: Updating base image versions (#1285)
balena-io/docs: pages/reference/base-images/base-images.md

## Features Overview

- Multiple Architectures:
  - armv5e
  - armv6
  - armv7hf
  - aarch64
  - amd64
  - i386
- Multiple Distributions:

>   - [Debian](https://www.debian.org/): jessie, stretch, buster and sid

Reviewer: I wonder if it's worth putting the Debian version numbers in brackets, similar to Ubuntu below?

- For each combination of distro and stack, we have two variants called `run` and `build`. The build variant is much heavier as it has a number of tools preinstalled to help with building source code. You can see an example of the tools that are included in the Debian Stretch variant [here]({{ $links.githubLibrary }}/base-images/blob/master/balena-base-images/armv7hf/debian/stretch/build/Dockerfile). The `run` variants are stripped down and only include a few useful runtime tools, see an example [here]({{ $links.githubLibrary }}/base-images/blob/master/balena-base-images/armv7hf/debian/stretch/run/Dockerfile). If no variant is specified, the image defaults to `run`
- The last optional field on tags is the date tag `<yyyymmdd>`. This is useful for production deployments as these base images are non-moving tags, so no packages in these will update ever.

#### Examples

`balenalib/raspberrypi3-node:10.17`

- `<hw>` : raspberrypi3 - The Raspberry Pi 3 device type.
- `<distro>` : omitted, so it defaults to Debian.
- `<lang>` : node - the Node.js runtime and npm will be installed

> - `<lang_ver>` : 10.17 - This gives us Node.js version 10.18.x whatever is the latest patch version provided on balenalib

Reviewer: Huh? Mix of 10.17 and 10.18

Author: Nice catch, which is why I want to autogenerate this as much as possible 😄

### How the Image Naming Scheme Works

With over 26000 `balenalib` base images to choose from, it can be overwhelming to decide which image and tag are correct for your project. To pick the correct image, it helps to understand how the images are named as that indicates what is installed in the image. In general, the naming scheme for the `balenalib` image set follows the pattern below:

```

> balenalib/<hw>-<distro>-<lang_stack>:<lang_ver>-<distro_ver>-(build|run)-<yyyymmdd>

Reviewer: Is it possible to specify `hw`, `distro`, `lang_stack` and `distro_ver` but **not** `lang_ver`?  Or is `lang_ver` compulsory if `lang_stack` is used?

Author: Yes as no tags are compulsory. Will default to `latest` if no tag set. Tested via:

```
docker run -it balenalib/raspberrypi3-debian-golang:stretch-run bash
root@ed271cd90107:/go# go version
go version go1.13.4 linux/arm

root@ed271cd90107:/go# cat/etc-os-release
bash: cat/etc-os-release: No such file or directory
root@ed271cd90107:/go# cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 9 (stretch)"
NAME="Debian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
VERSION_CODENAME=stretch
```

Reviewer: Ah cool! I only asked because it's not clear (from these docs) how the 'punctuation' (the hyphens and colons) would 'collapse' if you specified `hw`, `distro`, `lang_stack` and `distro_ver` but not `lang_ver`'

Could there ever be an edge-case with lang_ver and distro_ver (for the chosen lang_stack and distro) both being numeric, leading to a 'clash' between `balenalib/<hw>-<distro>-<lang_stack>:<lang_ver>` and `balenalib/<hw>-<distro>-<lang_stack>:<distro_ver>` ? (IYSWIM)
Although I'm sure all you clever folks already have a solution for that.

Author: Interesting Alpine and Python would fit the bill. I'm guessing there is a simple precedence

`docker run -it balenalib/raspberrypi3-alpine-python:3.7 bash` 

So this gives me Python 3.7 and the latest Alpine (`PRETTY_NAME="Alpine Linux v3.10"`). This makes intuitive sense (to me anyway 😄 ) given the lang ver is the first to be specified. If you wanted Alpine 3.7 and latest Python you'd have to do `docker run -it balenalib/raspberrypi3-alpine-python:latest-3.7 bash` which indeed gives Python 2.7.16 and `PRETTY_NAME="Alpine Linux v3.7"`

Author: And I guess rather than trying to explain that in text additional examples would be a good idea.

Reviewer: And I guess (for completeness) `balenalib/raspberrypi3-alpine-python:3.7-3.7` would give you Python3.7 on Alpine3.7?
Does the distro_ver also support a `latest` tag (i.e. `balenalib/raspberrypi3-alpine-python:3.7-latest`) or does `latest` only work for lang_ver ?

Author: Yes to the first question, 3.7 for both.

```
docker run -it balenalib/raspberrypi3-alpine-python:3.7-3.7 bash
bash-4.4# python --version
Python 3.7.4
bash-4.4# cat /etc/os-release
NAME="Alpine Linux"
ID=alpine
VERSION_ID=3.7.3
PRETTY_NAME="Alpine Linux v3.7"
HOME_URL="http://alpinelinux.org"
BUG_REPORT_URL="http://bugs.alpinelinux.org"
```

Latest also works with distro so your example as expected gives Python 3.7 and latest Alpine 3.10.3

```

#### Image Names

- `<hw>` is either architecture or device type and is **mandatory**. If using Dockerfile.templates, you can replace this with `%%BALENA_MACHINE_NAME%%` or `%%BALENA_ARCH%%`. For a list of available device names and architectures, see the [Device types](/reference/base-images/devicetypes/).
- `<distro>` is the Linux distribution. Currently there are 4 distributions, namely Debian, Alpine, Ubuntu and Fedora. This field is optional and will default to Debian if left out.

> - `<lang_stack>` is the programming language pack, currently we support Node.js, Python, OpenJDK and Go. This field is optional, and if left out, no language pack will be installed, so you will just have the distribution.

Reviewer: So I guess if you specify a `<lang_ver>` but not a `<lang_stack>`, the lang_ver will just be ignored?

Author: It won't be ignored and it'd probably just mean the image doesn't exist (unless it happened to coincide with an available tag) - but either way, it's unlikely to be what you want. For example, you can see the `raspberrypi3-alpine` tags here so if you specified a <lang_ver> tag of 10.8 (e.g. node) then there simply isn't a corresponding image.

Reviewer: Cool, I was just checking what the edge-case behaviour was :slightly_smiling_face:

Reviewer: Hmmm, that's curious - https://hub.docker.com/layers/balenalib/raspberrypi3-alpine/latest/images/sha256-054582a87ef96ba2243334500bb580246dba066967e9f6ad0c3bee655fde281d says "LABEL io.balena.device-type=raspberry-pi2" ??

#### Image Tags

In the tags, all of the fields are optional, and if they are left out, they will default to their `latest` pointer.

- `<lang_ver>` is the version of the language stack, for example, Node.js 10.10, it can also be substituted for `latest`.

> - `<distro_ver>` is the version of the Linux distro, for example in the case of Debian, there are 4 valid versions, namely `sid`, `jessie`, `buster` and `stretch`.

Reviewer: Should this mention that `latest` can also be used for the `<distro_ver>` tags?

- For each combination of distro and stack, we have two variants called `run` and `build`. The build variant is much heavier as it has a number of tools preinstalled to help with building source code. You can see an example of the tools that are included in the Debian Stretch variant [here]({{ $links.githubLibrary }}/base-images/blob/master/balena-base-images/armv7hf/debian/stretch/build/Dockerfile). The `run` variants are stripped down and only include a few useful runtime tools, see an example [here]({{ $links.githubLibrary }}/base-images/blob/master/balena-base-images/armv7hf/debian/stretch/run/Dockerfile). If no variant is specified, the image defaults to `run`
- The last optional field on tags is the date tag `<yyyymmdd>`. This is useful for production deployments as these base images are non-moving tags, so no packages in these will update ever.

#### Examples

`balenalib/raspberrypi3-node:10.18`

- `<hw>` : raspberrypi3 - The Raspberry Pi 3 device type.
- `<distro>` : omitted, so it defaults to Debian.
- `<lang>` : node - the Node.js runtime and npm will be installed

> - `<lang_ver>` : 10.18 - This gives us Node.js version 10.18.x whatever is the latest patch version provided on balenalib

Reviewer: The "Multiple language stacks:" list further up the page lists "Node.js 10.18.0" but here you're using a `<lang_ver>` of `10.18`. Does that imply that `balenalib/raspberrypi3-node:10.18`, `balenalib/raspberrypi3-node:10.18.0` and `balenalib/raspberrypi3-node:10` are all equally valid base-image names? Or is there some other logic at work?

Possibly worthy of more explanation? :man_shrugging:

Author: Yes. You can confirm this via `docker pull balenalib/raspberrypi3-node:10.18` and inspecting the output of `docker images` which shows they are all the same image (for this specific example). As I noted earlier I think a few more examples would go a long way for this sort of information.

```
docker images
REPOSITORY                                                         TAG                   IMAGE ID            CREATED             SIZE
balenalib/raspberrypi3-node                                        10                    3bf113fa9176        15 hours ago        231MB
balenalib/raspberrypi3-node                                        10.18                 3bf113fa9176        15 hours ago        231MB
balenalib/raspberrypi3-node                                        10.18.0               3bf113fa9176        15 hours ago        231MB
```

Reviewer: Hmmm, I wonder if this _might_ explain the behaviour you discovered at https://github.com/balena-io/docs/pull/1285#discussion_r362953784 where `balenalib/raspberrypi4-64-alpine-python:3-3.11-run` points to an image containing Python 3.6.9 (whereas you'd expect it to contain Python 3.8.0) ...

**_Perhaps_** when `python:3.6.9` got added, it also added the `python:3.6` and `python:3` tags, inadvertently "overwriting" the `python:3` tag which was previously pointing to `python:3.8.0`?

That's just a guess though, and only @nghiant2710 will know for sure.

	- i386
- Multiple Distributions:
	- [Debian](https://www.debian.org/): jessie, sid, stretch and buster
	- [Alpine](https://alpinelinux.org/): 3.6, 3.7, 3.8, 3.9, 3.10 and edge
	- [Ubuntu](https://www.ubuntu.com/): xenial (16.04), bionic (18.04), cosmic (18.10), disco (19.04) and eoan (19.10)
	- [Fedora](https://getfedora.org/): 26, 28, 29, 30 and 31
- Multiple language stacks:
	- [Node.js](https://nodejs.org/en/): 13.1.0, 12.13.0, 11.15.0, 10.17.0, 9.11.2, 8.16.1 and 8.11.1

> 	- [Python](https://www.python.org/): 2.7.17, 3.8.0, 3.7.5, 3.6.9, 3.5.7 and 3.4.10

Reviewer: 2.7.x, 3.8.x, 3.7.x, 3.6.x, 3.5.x and 3.4.x looks like a strange version-ordering system? :wink: 

Also, it seems some version-lists here are sorted largest -> smallest (e.g. Node.js), and others are sorted smallest -> largest (e.g. openJDK) ?

#### Examples

`balenalib/raspberrypi3-node:10.17`

- `<hw>` : raspberrypi3 - The raspberry pi 3 device type.
- `<distro>` : omitted, so it defaults to Debian.
- `<lang>` : node - the Node.js runtime and npm will be installed
- `<lang_ver>` : 10.17 - This gives us node.js version 10.17.x whatever is the latest patch version provided on balenalib
- `<distro_ver>` : omitted, so it defaults to `stretch`
- `(build|run)` : omitted, so the image defaults to the slimmed down `run` variant
- `<yyyymmdd>` : omitted, we don’t have a date frozen image, so new updates pushed to our 10.17 tag, for example patch versions from node.js will automatically be inherited when they are available.

`balenalib/i386-ubuntu-python:latest-bionic-build-20191029`

- `<hw>` : i386 - the intel 32 bit architecture that runs on Intel Edison
- `<distro>` : ubuntu
- `<lang>` : python

> - `<lang_ver>` : `latest` points to the latest python 2 version, which currently is 2.7.17

Reviewer 2: ```suggestion
- `<lang_ver>` : `latest` points to the latest Python 2 version, which currently is 2.7.17
```

- `<distro_ver>` : bionic is Ubuntu 18.04
- `(build|run)` : `build` - to include things like `build-essential` and `gcc`

> - `<yyyymmdd>` : 20191029 is a date frozen image - so this image will never be updated on dockerhub. Pinning to a date frozen base image is a good idea if you are running a fleet in production and are sensitive to dependencies updating and/or bandwidth constrained.

Reviewer 2: ```suggestion
- `<yyyymmdd>` : 20191029 is a date frozen image - so this image will never be updated on Docker Hub. Pinning to a date frozen base image is a good idea if you are running a fleet in production and are sensitive to dependencies updating and/or bandwidth constrained.
```

- `<hw>` : raspberrypi3 - The Raspberry Pi 3 device type.
- `<distro>` : omitted, so it defaults to Debian.
- `<lang>` : node - the Node.js runtime and npm will be installed

> - `<lang_ver>` : 10.18 - This gives us Node.js version 10.18.x whatever is the latest patch version provided on balenalib

Reviewer: The 10.18 here is now fixed, but the example image-name a few lines up is still `balenalib/raspberrypi3-node:10.17` ? :eyes:

### Supported Architectures, Distros and Languages

Currently, balenalib supports the following OS distributions and Language stacks, if you would like to see others added, create an issue on the [balena base images repo]({{ $links.githubLibrary }}/base-images/issues).

> | Distribution | Default                      | Supported Architectures                      |

Reviewer: Presumably in both this table and the next, the "Default" column is the version you'd get if you used the `latest` tag for the `<distro_ver>` or `<lang_ver>` fields?
Is it worth mentioning that explicitly? (e.g. by changing the "Default" column-header to "Default (latest)" ? ) :man_shrugging: 

---

Hmmm, in the "Multiple Distributions:" and "Multiple language stacks:" lists earlier on in the page, I wonder if it'd be worth highlighting the versions corresponding to the "latest" tag by e.g. displaying them in bold? Like:
 - [Node.js](https://nodejs.org/en/): 8.11.1, 8.17.0, 9.11.2, 10.18.0, 11.15.0, 12.14.0 and **13.5.0**
 - [Python](https://www.python.org/): **2.7.17**, 3.4.10, 3.5.7, 3.6.9, 3.7.5 and 3.8.0

| ------------ | ---------------------------- | -------------------------------------------- |
| Debian       | Debian GNU/Linux 10 (buster) | armv5e, armv6, armv7hf, aarch64, amd64, i386 |

> | Alpine       | Alpine Linux v3.10           | armv6, armv7hf, aarch64, amd64, i386         |

Reviewer: Hmmm, I've just noticed that this table says that the Default version of Alpine is v3.10, and by comparing Digests on https://hub.docker.com/r/balenalib/raspberrypi3-alpine/tags you can also confirm that the `latest` and `3.10` tags point to the same image, and yet Alpine 3.11 is mentioned in the "Multiple Distributions:" list on this page, and there's a `3.11` tag on DockerHub? :confused:

Reviewer: _Maybe_ that's explained by https://github.com/balena-io/contracts/commit/84035e0310ef5f14fe3fd956986091084e5ca680#r36690619 but that's a bit of a random stab in the dark :wink:

Author: Yeah looks like a mistake as mentioned in that linked contract above. In general though `latest` doesn't necessarily mean "latest" 😕https://medium.com/@mccode/the-misunderstood-docker-tag-latest-af3babfd6375  In this case though it looks like it should have been tagged latest.

Reviewer: That's interesting, but I've no idea how and/or if Docker tags relate to the base-image tags used by Balena :wink:

Author: I'm not sure I understand that comment. The tags on Docker Hub are the base-image tags.

Reviewer: The medium.com article you've linked to mentions tags like `marc/test:latest` and it seems like the `latest` tag there is automatically added by Docker. I was assuming that that's actually unrelated to the `latest` tag embedded within a Balena image-name like `balenalib/raspberrypi3-alpine-python:latest-3.7` ? :man_shrugging: :man_shrugging:

| Ubuntu       | 18.04 LTS (bionic)           | armv7hf, aarch64, amd64, i386                |
| Fedora       | Fedora 30                    | armv7hf, aarch64, amd64, i386                |

| Language | Default | Supported Architectures                      |
| -------- | ------- | -------------------------------------------- |
| Node.js  | v13.5.0 | armv6, armv7hf, aarch64, amd64, i386         |
| Python   | v2.7.17 | armv5e, armv6, armv7hf, aarch64, amd64, i386 |
| OpenJDK  | v11-jdk | armv7hf, aarch64, amd64, i386, armv6         |
| Go       | 1.13.4  | armv7hf, aarch64, amd64, i386, armv6         |

> | Dotnet   | 3.1-sdk | armv7hf, aarch64, amd64                      |

Reviewer: Some of the versions here have a `v` at the front and some don't? I guess that'll be fixed when this page is more auto-generated? :slightly_smiling_face:

Author: Thanks for these nits as good to clean up. I've removed `v` as clearly these shouldn't be specified in the tag and could lead to confusion.

#### Notes

> Devices with a [device type](/reference/base-images/devicetypes/) of `raspberry-pi` (Raspberry Pi1 and Zero) will be built from `balenalib/rpi-raspbian` and will be [Raspbian]({{ $links.githubLibrary }}/resin-rpi-raspbian) base images. Raspberry Pi2 and Pi3 base images have the Raspbian package source added and Raspbian userland pre-installed.

Reviewer: "Raspberry Pi2 and Pi3 base images have the Raspbian package source added and Raspbian userland pre-installed."

Is this true for just Debian on Raspberry Pi, or for the other OSes too? (I guess it wouldn't make sense to add the Raspbian package source to Fedora?)

Also, I think "Raspberry Pi2 and Pi3" ought to be "Raspberry Pi2 and Raspberry Pi3" ? :man_shrugging:

Reviewer: ...and there's also the Raspberry Pi 4 of course :wink: Maybe "Raspberry Pi2 and above" ?

Author: No for Pi4, as we only offer 64-bit support (and I need to explicitly mention that for Pi3 variant).  I'm checking the distros and looks like only Debian adds Raspbian package source via https://github.com/balena-io-library/base-images/blob/master/balena-base-images/device-base/raspberrypi3/debian/jessie/build/Dockerfile. Maybe @nghiant2710 can confirm this though before I update.

Author: This issue is also related as this doesn't work for `debian:sid` variant https://github.com/balena-io-library/base-images/issues/602

Reviewer: Interesting - why no Raspbian on RPi4 on balena, when the images at https://www.raspberrypi.org/downloads/raspbian/ support the RPi4?

Author: There is no balena `raspberrypi4` device unlike the Pi3, there is only `raspberrypi4-64`.

Not all OS distro and language stack versions are compatible with each other. Notice that there are some combinations that are not available in the `balenalib` base images.

- [Nodejs dropped 32-bit builds](https://github.com/nodejs/build/issues/885) a while ago so i386-based nodejs images (Debian, Fedora and Ubuntu) v8.x and v6.x are official. New series (v10.x and v12.x) are using unofficial builds so please notice when using these images.

> - armv6 binaries were officially dropped from v12 and v12 armv6 support is now considered unofficial.

Reviewer: Is the `v12` on this line referring to Nodejs?

> - The Nodejs v6.x and v8.x series are not available for i386 Alpine Linux base images v3.9 and edge as node crashes with segfault error, we are investigating the issue and will add them back as soon as the issue is resolved.

Reviewer: Nitpick: Nodejs v6.x isn't included in the list of supported Node.js versions earlier on the page :wink:

Author: Thanks, I left this as while it isn't supported there are v6 base images available on Docker Hub.

## Installing Packages

Installing software packages in balenalib containers is very easy, and in most cases, you can just use the base image operating system package manager. However to make things even easier, every balenalib image includes a small `install_packages` script that abstracts away the specifics of the underlying package managers, and adds the following useful features:

- Install the named packages, skipping prompts etc.
- Clean up the package manager metadata afterward to keep the resulting image small.
- Retries if package install fails. Sometimes a package will fail to download due to a network issue, and retrying may fix this, which is particularly useful in an automated build pipeline.

An example of this in action is as follows:

```Dockerfile
FROM balenalib/raspberrypi3

RUN install_packages wget git

CMD ["bash", "start.sh"]
```

> This will run an `apt-get update -qq`, then install `wget` and `git` via apt-get with `-y --no-install-recommends` flags, and it will by default try this 2 times before failing. You can see the source of `install_packages` [here]({{ $links.githubLibrary }}/base-images/blob/master/balena-base-images/armv7hf/debian/stretch/run/Dockerfile#L25-L49).

Reviewer: Should this be `#L26-L49` ?

At the end of a container's lifecycle, when a request to container restart, reboot or shutdown is sent to the supervisor, the [balenaEngine](https://www.balena.io/engine/) will send a `SIGTERM` (signal 15) to the containers, and 10 seconds later it will issue a `SIGKILL` if the container is still running. This timeout can also be configured via the [stop_grace_period](https://docs.docker.com/compose/compose-file/compose-file-v2/#stop_grace_period) in your docker-compose.yml.

## Working with Dynamically Plugged Devices

In many IoT projects, your containers will want to interact with some hardware, often this hardware is plugged in at runtime, in the case of USB or serial devices. In these cases, you will want to enable [`udevd`][udevd-link] in your container. In `balenalib` images this can easily be done either by adding `ENV UDEV=1` in your `Dockerfile` or by setting an [environment variable](https://www.balena.io/docs/learn/manage/serv-vars/).

> You will also need to run your container `privileged`, by default. Any [balenaCloud](https://www.balena.io/cloud/) projects that don't contain a `docker-compose.yml` will run their containers `privileged`. If you are using a multicontainer project, you will need to add `privileged: true` to each of the service definitions for the services that need hardware access.

Reviewer: Was the full-stop put in the wrong place here? Looks like maybe it should say "You will also need to run your container `privileged`. By default, any ..." ?

| Distribution | Default                | Supported Architectures                      |
|---------|------------------------------|----------------------------------------------|
| Debian  | Debian GNU/Linux 10 (Buster) | armv5e, armv6, armv7hf, aarch64, amd64, i386 |
| Alpine  | Alpine Linux v3.10            | armv6, armv7hf, aarch64, amd64, i386 		|
| Ubuntu  | 18.04 LTS (Bionic Beaver)    | armv7hf, aarch64, amd64, i386                |

> | Fedora  | Fedora 30 (Thirty)      | armv7hf, aarch64, amd64, i386                |

Reviewer 2: ```suggestion
| Fedora  | Fedora 30               | armv7hf, aarch64, amd64, i386                |
```

| Language | Default  	                  | Supported Architectures                      |
|---------|------------------------------|----------------------------------------------|
| Node.js | v13.1.0                      | armv6, armv7hf, aarch64, amd64, i386         |

> | Python  | v2.7.17                       | armv5e, armv6, armv7hf, aarch64, amd64, i386 |

Reviewer: I wonder if it's also worth having a separate Python3 language tag, bearing in mind the upcoming EOL of Python2 ? https://pythonclock.org/

Reviewer: There's also part of the page which says "`latest` points to the latest Python 2 version, which currently is 2.7.17" and I guess that means there's currently no way to automatically get the latest Python 3 version?

Author: I missed this one. I'd have thought using `3` would do it but that doesn't seem to work as expected:

```
docker run -it balenalib/raspberrypi4-64-alpine-python:3-3.11-run bash
Python 3.6.9
```

Author: Maybe that's the default package available i.e via `apt-get install python3`?

Reviewer: Created https://github.com/balena-io-library/base-images/issues/606 to track this inconsistency

Not all OS distro and language stack versions are compatible with each other. Notice that there are some combinations that are not available in the `balenalib` base images.

> - [Node.js dropped 32-bit builds](https://github.com/nodejs/build/issues/885) a while ago so i386-based nodejs images (Debian, Fedora and Ubuntu) v8.x and v6.x are official. New series (v10.x and v12.x) are using unofficial builds so please notice when using these images.

Reviewer: What does "so please notice when using these images" mean? :wink:
